streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -481,14 +486,16 @@ int commitAll() {
         if (rebalanceInProgress) {
             return -1;
         } else {
-            int commits = 0;
+            final Map<TaskId, Map<TopicPartition, OffsetAndMetadata>> consumedOffsetsAndMetadataPerTask = new HashMap<>();
             for (final Task task : tasks.values()) {
                 if (task.commitNeeded()) {
-                    task.commit();
-                    commits++;
+                    consumedOffsetsAndMetadataPerTask.put(task.id(), task.prepareCommit());
+                    task.markCommitted();
```

Comment:
We could also do a second loop over all tasks, _after_ calling `commit(..)` below -- not sure if this is ok as-is?